src/membership/cut_detector.rs

```diff
@@ -0,0 +1,162 @@
+use crate::{
+    common::Endpoint,
+    transport::proto::{AlertMessage, EdgeStatus},
+};
+
+use std::collection::{HashMap, HashSet};
+
+const NUM_MIN: usize = 3;
+
+pub struct MultiNodeCutDetector {
+    /// Number of observers per subject and vice versa
+    num: usize,
+    /// High water mark
+    high: usize,
+    /// Low water mark
+    low: usize,
+    proposal_count: usize,
+    updates_in_progress: usize,
+    reports_per_host: HashMap<Endpoint, HashMap<usize, Endpoint>>,
+    proposal: Vec<Endpoint>,
+    pre_proposal: HashSet<Endpoint>,
+    seen_down_link_events: bool,
+}
+
+impl MultiNodeCutDetector {
+    pub fn new(num: usize, high: usize, low: usize) -> Self {
+        if high > num || low > high || num < NUM_MIN || low <= 0 || high <= 0 {
+            panic!("Invalid arguments passed to cut detector")
+        }
+
+        Self {
+            num,
+            high,
+            low,
+            proposal_count: 0,
+            updates_in_progress: 0,
+            reports_per_host: HashMap::new(),
+            proposal: Vec::new(),
+            pre_proposal: HashSet::new(),
+            seen_down_link_events: false,
+        }
+    }
+
+    fn aggregate(message: AlertMessage) -> Vec<Endpoint> {
+        msg.ring_number
+            .iter()
+            .map(|ring_number| {
+                self.aggregate_for_proposal(
+                    message.edge_src,
+                    message.edge_dst,
+                    msg.edge_status,
+                    ringN,
+                )
+            })
+            .flatten()
+            .collect()
+    }
+
+    fn aggregate_for_proposal(
+        &mut self,
+        link_src: Endpoint,
+        link_dst: Endpoint,
+        edge_status: EdgeStatus,
+        ring_number: usize,
+    ) -> Vec<Endpoint> {
+        debug_assert!(ring_number <= self.num);
+
+        if edge_status == EdgeStatus::Down {
+            self.seen_down_link_events = true;
+        }
+
+        let reports_per_host = self
+            .reports_per_host
+            .entry(link_dst)
+            .or_insert(HashMap::new());
+
+        if (reports_per_host.contains(ring_number)) {
+            // Duplicate announcement, ignore
+            return vec![];
```

Comment:
```suggestion
            return Vec::new();
```
just personal pref :D